lib/Storage/S3StorageClient.cpp

```diff
@@ -0,0 +1,379 @@
+/// \file S3StorageClient.cpp
+/// \brief Implementation of StorageClient operations in S3
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <fstream>
+
+#include "llvm/Support/Process.h"
+#include "llvm/Support/YAMLTraits.h"
+
+#include "revng/Storage/Path.h"
+#include "revng/Storage/S3StorageClient.h"
+#include "revng/Support/Assert.h"
+#include "revng/Support/OnQuit.h"
+#include "revng/Support/PathList.h"
+
+#include "aws/core/Aws.h"
+#include "aws/core/auth/AWSCredentials.h"
+#include "aws/core/utils/logging/FormattedLogSystem.h"
+#include "aws/s3/S3Client.h"
+#include "aws/s3/model/CopyObjectRequest.h"
+#include "aws/s3/model/DeleteObjectRequest.h"
+#include "aws/s3/model/GetObjectRequest.h"
+#include "aws/s3/model/HeadObjectRequest.h"
+#include "aws/s3/model/PutObjectRequest.h"
+
+using FilenameMapType = llvm::StringMap<std::string>;
+
+template<>
+struct llvm::yaml::CustomMappingTraits<FilenameMapType> {
+  static void inputOne(IO &IO, llvm::StringRef Key, FilenameMapType &Value) {
+    IO.mapRequired(Key.str().c_str(), Value[Key]);
+  }
+
+  static void output(IO &IO, FilenameMapType &Value) {
+    for (auto &Pair : Value) {
+      IO.mapRequired(Pair.first().str().c_str(), Pair.second);
+    }
+  }
+};
+
+namespace revng::detail {
+
+namespace {
+using Aws::Utils::Logging::FormattedLogSystem;
+using Aws::Utils::Logging::LogLevel;
+
+static Logger<> AWSLogger("aws");
+
+class LoggerSystem : public FormattedLogSystem {
+private:
+  std::mutex Mutex;
+
+public:
+  LoggerSystem(LogLevel LogLevel) : FormattedLogSystem(LogLevel) {}
+  ~LoggerSystem() override = default;
+  void Flush() override { AWSLogger.flush(); }
+
+  void ProcessFormattedStatement(Aws::String &&Statement) override {
+    // This function needs to be thread-safe, hence the lock_guard
+    std::lock_guard Guard(Mutex);
+    revng_log(AWSLogger, Statement);
+  }
+};
+
+class SDKInitImpl {
```

Comment:
Using a `class` for this kind of things makes sense if you use the destructor.

You could turn `std::optional<SDKInitImpl> SDKInit` into a simple `bool SDKIsInitialized = false` and copy the body of the constructor where you're currently checking the optional.

If you go this route, you also probably need to capture `Options` by copy or make a global variable out of it.

Give it a think.

Reply:
Changed to:
```cpp
bool SDKIsInitialized = false;
void initializeSDK(); // asserts if called twice
```